### PR TITLE
First draft for state retention

### PR DESCRIPTION
**Pull request by TCke83**
First draft without tests for keeping track of the state of a SwitchControl, AnalogInfoControl & DigitalInfoControl and update the state through the corresponding state class.

If this gets approved I will need some help with the kotlin and/or groovy tests, not really familiar with those.

### CODE REVIEW

File: src/main/java/cz/smarteon/loxone/message/EncryptedCommand.java [hunk 1]
```diff
@@ -102,10 +102,6 @@ public static EncryptedCommand<Token> refreshToken(final String tokenHash, final
     }
 
     private static String encodeUrl(String toEncode) {
-        try {
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/message/EncryptedCommand.java:105]: could you kindly move changes in this file to separate commit?

File: src/main/java/cz/smarteon/loxone/message/TextEvent.java [hunk 2]
```diff
@@ -1,15 +1,27 @@
...
 import static java.util.Objects.requireNonNull;
 
 /**
  * Loxone event carrying text value.
  */
+@Getter
+@ToString(callSuper = true)
 public class TextEvent extends LoxoneEvent {
+
+    /**
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/message/TextEvent.java:17]: does lombok somehow move this to getters? Otherwise it makes no sense to document it - prvate members are usually not visible in api doc

File: src/main/java/cz/smarteon/loxone/app/state/events/LockedEvent.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+@Value
+@Builder
+@Jacksonized
+public class LockedEvent {
+
+    @Builder.Default
+    Locked locked = Locked.NO;
+
+    String reason;
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/events/LockedEvent.java:16]: should be private?

File: src/main/java/cz/smarteon/loxone/app/state/ControlState.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+     * The control that this state refers to.
+     */
+    protected final T control;
+
+    /**
+     * Method that accepts ValueEvent from the miniserver to update the internal state.
+     * @param event value event received (should not be null)
+     */
+    void accept(@NotNull ValueEvent event) {
+        // default implementation
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/ControlState.java:32]: does the default impl. make any sense? Isn't it better to have the method abstract?

File: src/main/java/cz/smarteon/loxone/app/state/DigitalState.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package cz.smarteon.loxone.app.state;
+
+/**
+ * Enum containing the different switch states for following controls.
+ * <ul>
+ *     <li><code>SwitchControl</code></li>
+ *     <li><code>DigitalInfoControl</code></li>
+ * </ul>
+ */
+public enum DigitalState {
```
Comment (jimirocks) [inline src/main/java/cz/smarteon/loxone/app/state/DigitalState.java:10]: Well, I still see little value in this class. In case of analog there is a Double which is nul when not initialized or having a value when initialized. Why in case of digital there can't be `Boolean` which is null when not initialized.

Or vice versa, why there is no corresponding AnalogState class.

Please consider to have this consistent.

**Review by jimirocks — CHANGES_REQUESTED** [review]
Thank you for the effort, please consider comments.

**But the most important thing:** Let's not add `LoxoneState` into `Loxone`. Since State depends on Loxone having it dependent also the opposite way is rather unpractical and coudl bring significant problems in future development. I strongly prefer having `LoxoneState` as separate additional layer on top of `Loxone` class. **Does the concept make sense to you?**

**As for unit tests:** 
- Recently we migrated from groovy - so no need to learn groovy.
- we prefer [Junit5 with Kotlin](https://www.baeldung.com/kotlin/junit-5-kotlin), [Strikt](https://strikt.io/) and [Mockk](https://mockk.io/) - I will add some guide later
- but also using Junit5 + Java would be acceptable

File: src/main/java/cz/smarteon/loxone/message/TextEvent.java (hunk 2 above)
Comment (TCke83) [reply src/main/java/cz/smarteon/loxone/message/TextEvent.java:17]: Documentation on fields is copied to the getters and setters at compile time.

File: src/main/java/cz/smarteon/loxone/app/state/events/LockedEvent.java (hunk 3 above)
Comment (TCke83) [reply src/main/java/cz/smarteon/loxone/app/state/events/LockedEvent.java:16]: The @Value annotation marks the fields private at compile time, i could add it to the source code, but it is not needed because lombok adds it

File: src/main/java/cz/smarteon/loxone/message/EncryptedCommand.java (hunk 1 above)
Comment (TCke83) [reply src/main/java/cz/smarteon/loxone/message/EncryptedCommand.java:105]: Added in separate pull request

File: src/main/java/cz/smarteon/loxone/app/state/ControlState.java (hunk 4 above)
Comment (TCke83) [reply src/main/java/cz/smarteon/loxone/app/state/ControlState.java:32]: I assumed the state classes would not always be interested in TextEvents & ValueEvents and to avoid an empty accept method in the implementation i've provided the default implementation (doing nothing with the event) in the abstract class.

File: src/main/java/cz/smarteon/loxone/app/state/DigitalState.java (hunk 5 above)
Comment (TCke83) [reply src/main/java/cz/smarteon/loxone/app/state/DigitalState.java:10]: Okay makes sense i've used a Boolean to align it with the AnalogState

**Comment by TCke83** [comment]
Going to look further into avoiding inclusion of LoxoneState class in Loxone class later on, it makes sense not adding it to the Loxone base class.

**Comment by TCke83** [comment]
Added LoxoneState as wrapper around the Loxone class for state retention. Please take a look if this is the idea you had in mind?

Maybe we can deprecated the two constructors with params in the Loxone class in favor of the constructor with the LoxoneProfile parameter.

**Comment by jimirocks** [comment]
Hi, thanks for comments and changes, now it looks acceptable once there are some unit tests. So are you able to add them?

I'll think about deprecation of the two former Loxone constructors, but I feel they still have some value in matter of simplicity...

**Comment by jimirocks** [comment]
Hello @TCke83 do you still plan to finish this?

**Comment by TCke83** [comment]
> Hello @TCke83 do you still plan to finish this?

Hello, yes i was planning to last week, but some unexpected things came up. Currently i'm on holiday, Will try to finish the unit tests by the end of the month.

**Comment by jimirocks** [comment]
@TCke83 any news?

**Comment by TCke83** [comment]
Still didn't found the time, sorry for the delay, still planning on adding some tests for the new code. Will do my best to try to do it as soon as possible.

**Comment by jimirocks** [comment]
well we are not in hurry from ourpoint of view - just wanted to know if there is still plan to finish it ;-)
